### DISCUSSION TITLE
Storage of CI artifacts for testing

### DISCUSSION THREAD

**Discussion by oursland** [FreeCAD/FreeCAD#17645] [Code contribution discussions]
From the Developer's Meeting

* Developers and users have requested the ability to download the build artifacts from CI for testing changes locally. Currently, these artifacts are not being preserved as they would consume the valuable storage budget.
* With the increase in requests for this functionality, perhaps it is time to consider requesting FPA funds to increase storage capabilities and adding the temporary preservation of build artifacts.

## Documentation

* [Storing and sharing data from a workflow](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/storing-and-sharing-data-from-a-workflow)
    * By default, GitHub stores build logs and artifacts for 90 days, and this retention period can be customized. For more information, see "[Usage limits, billing, and administration](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#artifact-and-log-retention-policy)."
* [Storing and sharing data from a workflow](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/storing-and-sharing-data-from-a-workflow)
* [Downloading workflow artifacts](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/downloading-workflow-artifacts)

## Pricing

According to the [GitHub Pricing Calculator](https://github.com/pricing/calculator#packages), additional package storage is $0.25 USD per GB, and additional data transfer out outside of Actions $0.50 USD per GB.

![image](https://github.com/user-attachments/assets/861079d5-4df1-4c39-8ee0-004c523a732c)

**Comment by oursland** [comment]
@adrianinsaval

**Comment by adrianinsaval** [comment]
I don't believe that github packages is the same as actions artifacts, from my understanding github packages are for specific formats such as docker images, npm and others, see https://docs.github.com/en/packages/working-with-a-github-packages-registry.

In the documentation for actions artifacts I don't see anywhere if there is any limit on how much storage we can use (as a public repo) for artifacts, it says you get 500mb for free but as I understand it this what you get for your private repos

> Storing artifacts uses storage space on GitHub. GitHub Actions usage is free for standard GitHub-hosted runners in public repositories, and for self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage for use with GitHub-hosted runners, depending on the account's plan. Any usage beyond the included amounts is controlled by spending limits. For more information, see "[Managing billing for GitHub Actions](https://docs.github.com/en/billing/managing-billing-for-github-actions)."

**Comment by oursland** [comment]
My use of GitHub Packages as a proxy is due to them using the packages icon for the artifacts.  If it is free for us, then there is no reason we should not simply add artifacts to each CI run.

The next issue would be how to provide these artifacts.  Users would expect that they are provided packages similar to those provided by the weekly builds.

The draft PR for [pixi](https://github.com/FreeCAD/FreeCAD/pull/16795) for developer environment management is intended to also support a `pixi run package` to generate packages.  Unfortunately, this PR has stalled by a build issue that is not understood.

**Reply by adrianinsaval** [reply]
a full appimage for every run is likely not a good idea, but we could make a script to recreate the ci conda environment on a local machine and install the freecad build on it.

**Reply by oursland** [reply]
It wouldn't just be the AppImage, it'd also be a .dmg, and .7z installer for testing by non-developers.  It is a fair bit of additional overhead.
